docs/rfc/rfc-010-p2p-light-client.rst

```diff
@@ -0,0 +1,127 @@
+forward, and the P2P-backed statesync implementation that took place during
+the 0.35 cycle. The light client only really needs to be able to request (and
+receive) a `LightBlock` at a given height. To support this, a new Reactor
+would run on every full node and validator which would be able to service
+these requests. The workload would be entirely request-response, and the
+implementation of the reactor would likely be very straight forward, and the
+implementation of the provider is similarly relatively simple. 
+
+The complexity of the project focuses around peer discovery, handling when
+peers disconnect from the light clients, and how to change the current P2P
+code to appropriately handle specialized nodes.
+
+I believe it's safe to assume that much of the current functionality of the
+current ``light`` mode would *not* need to be maintained: there is no need to
+proxy the RPC endpoints over the P2P layer and there may be no need to run a
+node/process for the p2p light client (e.g. all use of this will be as a
+client.) 
+
+The ability to run light clients using the RPC system will continue to be
+maintained.
+
+LibP2P
+~~~~~~
+
+While some aspects of the P2P light client implementation are orthogonal to
+LibP2P project, it's useful to think about the ways that these efforts may
+combine or interact.
+
+We expect to be able to leverage libp2p tools to provide some kind of service
+discovery for tendermint-based networks. This means that it will be possible
+for the p2p stack to easily identify specialized nodes, (e.g. light clients)
+thus obviating many of the design challenges with providing this feature in
+the context of the current stack.
+
+Similarly, libp2p makes it possible for project to be able back their non-Go
+light clients, without the major task of first implementing tendermint's p2p
+connection handling. We should identify if there exist users (e.g. the go IBC
+relayer, it's maintainers, and operators) who would be able to take advantage
+of p2p light client, before switching to libp2p. To our knowledge there are no
+external implementations of this p2p protocol, and it seems unlikely that a
+team would implement this directly ahead of its impending removal.
+
+User Cases
+~~~~~~~~~~
+
+This RFC makes a few assumptions about the use cases and users of light
+clients in tendermint.
+
+One of the most active use cases for light clients is in the implementation of
+the IBC relayer. Thus, we expect that providing P2P light clients might
+increase the reliability of relayers and reduce the cost of running a relayer,
+because relayer operators won't have to decide between rely on public RPC
+endpoints (unreliable) or running their own full nodes (expensive.) This
+also assumes that there are *no* other uses of the RPC in the relayer, and
+unless the relayers have the option of dropping all RPC use, it's unclear if a
+P2P light client will actually be able to successfully remove the dependency
+on the RPC system.
```

Comment:
Yeah so relayers still need to be able to query the application state (specifically the IBC module) for pending txs to transfer across. I know that the SDK are trying to migrate more to gRPC and thus it's possible that the only use of the RPC may be the "local" server.

Reply:
It might be that in a years time, no "external" client uses the Tendermint RPC because it's wrapped by the SDK's gRPC server. Node operators and developers might still use the Tendermint RPC for debugging and monitoring purposes and non-sdk chains will probably also still use the RPC. Does that provide a little extra context? 

For now, the IBC will still be using the Tendermint RPC not for the verification of headers but to read the application state, fetch the pending IBX transactions and then submitting those transactions on the destination chain.

Reply:
I guess that makes sense, though perhaps it makes more sense to have (for these cases) the light client implemented on top of the SDKs grpc end points rather than based on the p2p layer? 

If the goal is reduce operational complexity for relayers, then needing them to have fewer connections total, is perhaps the real solution.

Reply:
> the light client implemented on top of the SDKs grpc end points rather than based on the p2p layer?

This would make sense. Have a look at this: https://github.com/cosmos/cosmos-sdk/issues/11012